das/daser.go

```diff
@@ -90,5 +149,91 @@ func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
 		sampleTime := time.Since(startTime)
 		log.Infow("sampling successful", "height", h.Height, "hash", h.Hash(),
 			"square width", len(h.DAH.RowsRoots), "finished (s)", sampleTime.Seconds())
+
+		checkpoint = h.Height
 	}
 }
+
+// catchUpJob represents a catch-up job. (from:to]
+type catchUpJob struct {
+	from, to int64
+}
+
+// catchUpScheduler spawns and manages catch-up jobs, exiting only once all jobs
+// are complete and the last known DASing checkpoint has been stored to disk.
+func (d *DASer) catchUpScheduler(ctx context.Context, checkpoint int64) {
+	wg := sync.WaitGroup{}
+
+	defer func() {
+		// wait for all catch-up jobs to finish
+		wg.Wait()
+		// store latest DASed checkpoint to disk here to ensure that if DASer is not yet
+		// fully caught up to network head, it will resume DASing from this checkpoint
+		// up to current network head
+		// TODO @renaynay: Implement Share Cache #180 to ensure no duplicate DASing over same
+		//  header
+		if err := storeCheckpoint(d.cstore, checkpoint); err != nil {
+			log.Errorw("storing checkpoint to disk", "height", checkpoint, "err", err)
+		}
+		// signal that all catchUp routines have finished
+		close(d.catchUpDn)
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case job := <-d.jobsCh:
+			// spawn catchUp routine
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				d.catchUp(ctx, job)
+				// TODO @renaynay: assumption that each subsequent job is to a higher height than the previous.
+				//  I don't see why that is *not* the case, though.
+				checkpoint = job.to
+			}()
+		}
+	}
+}
+
+// catchUp starts a sampling routine for headers starting at the next header
+// after the `from` height and exits the loop once `to` is reached. (from:to]
+func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) {
+	routineStartTime := time.Now()
+	log.Infow("sampling past headers", "from", job.from, "to", job.to)
+
+	// start sampling from height at checkpoint+1 since the
+	// checkpoint height is DASed by broader sample routine
+	for height := job.from + 1; height <= job.to; height++ {
+		h, err := d.getter.GetByHeight(ctx, uint64(height))
+		if err != nil {
+			if err == context.Canceled {
+				return
+			}
+
+			log.Errorw("failed to get next header", "height", height, "err", err)
+			continue // TODO @renaynay: should we really continue in this case?
```

Comment:
cc @Wondertan

Reply:
If the error happens here then something is really wrong, so I would propose full return here, so node has to be restarted to start catching up here again.

Reply:
Alternative would be to retry the header a few times before giving up. But could be done in a separate PR.

Reply:
Although, there is another argument - is pruning from the future. If some headers were pruned, the daser should be ok with the `header.ErrNotFound`, but:
* It should be handled in a specifical case
* We should handle pruning in a way that won't involve ranging over each unavailable header. Meaning that DASer should be somehow aware of pruned headers, so he won't waste any resources on them.

So making a return here as default should be ok for the reason in the comment above.

Reply:
Despite pruning there is the assumption that all headers will be _somehow_ downloadable from the network. Note: this is only relevant for nodes that sync from genesis anyways. I think it is safe to assume that in between the the unbonding period headers are available without any further gymnastics (besides maybe asking a max a hand full of nodes).

Reply:
True. Right now, syncing from genesis is the only default, so if there are no headers, something is really wrong and stopping adding is the simplest default we can do tight now.

Regarding asking for headers if nothing is locally. I want this to be a case for our API, so if the pruning is on and someone asks a node for a very old header, it should take if from p2p network and provide to the api.

Reply:
tracked: https://github.com/celestiaorg/celestia-node/issues/547

i implemented a full return for now.
can i resolve this?

Reply:
Yes.

Reply:
But retries here are useless.

Reply:
See https://github.com/celestiaorg/celestia-node/issues/547#issuecomment-1077661008